design-proposals/macvtap-mode-for-pod-networking/macvtap-mode-for-pod-networking.md

```diff
@@ -0,0 +1,166 @@
+But **having a pod network is still mandatory for the many environments where VM is communicating with pods and Kubernetes services.**
+
+This change allows to bind the macvtap interface directly to veth, so the VM will get performance of native pod networking:
+
+```
+{CNI magic}  [veth]<--->[veth]
+                     ^
+                     |
+                     '-->[macvtap]<---->[eth0]
+^---------host--------^ ^----pod---^  ^--vm--^
+```
+
+As shown by benchmarks, it gets almost the same throughput as native veth interface:
+
+![flannel](latency-flannel.svg)
+![cilium](latency-cilium.svg)
+
+## Goals
+Provide a more efficient method to bind pod networking.
+
+## Non Goals
+Same as bridge the macvtap mode for pod networking does not support live-migration, unless the [`NetworkAwareLiveMigration`](https://github.com/kubevirt/community/pull/182) feature be merged.
+
+## Definition of Users
+Users who are looking for less latency mode to bind pod networking.
+
+## User Stories
+* As a user / admin, I want to bind a VM directly to a pod network without having overhead on additional network interfaces and masquerade.
+
+## Repos
+- [KubeVirt](https://github.com/kubevirt/kubevirt)
+
+# Design
+KubeVirt already has macvtap binding mode, but it is working only with [macvtap-cni](https://github.com/kubevirt/macvtap-cni).  
+This change is going to extend its logic to make it working with pod network and other CNIs as well.
+
+If `macvtap` mode requested, then the folowing check is invoked:
+
+- If the pod interface is **macvtap**, usually provided by macvtap-cni, then attach VM directly to it. This is left for compatibility with macvtap-cni.
+
+- If the pod interface is **veth** or any other type, then create a macvtap interface and bind to it.  
+  Due to the inability to use a downward veth interface for communicating with VM, you also have to prepare macvlan just to bind DHCP server to it.
+
+  The final scheme:  
+  ![](macvtap-mode-for-pod-networking.svg)
+
+Also this change extends existing macvtap driver to write cached information about the interface and prepare DHCP config.
+
+**How is the macvtap device created?**
+
+The new flag is added to the **tap-device-maker** binary. If it is called with `--parent-name` specified, then macvtap device will be created instead of tap.  
+But unlike tap, before running VM with macvtap interface, you need need to:
+
+- Set device rule for the cgroup running this pod:
+
+  *more details here: https://www.kernel.org/doc/Documentation/cgroup-v1/devices.txt*
+
+- Create character device in pod mount namespace:
+  ```
+  cat /sys/class/net/tap0/macvtap/tap5/dev
+  238:1
+  mknod /dev/tap5 c 238 1
+  ```
+
+## API Examples
+Thanks to macvtap-cni we already have everything needed and we don't need to modify the API.  
+From the user perspective we just allow to using macvtap mode to bind pod networking, example:
```

Comment:
maybe add more detail into how you will "allow using macvtap mode to bind pod networking" - i.e. [updating the admitter](https://github.com/kubevirt/kubevirt/blob/392ee0cae3e8477ca1c158984eed44faa8a184a3/pkg/virt-api/webhooks/validating-webhook/admitters/vmi-create-admitter.go#L89).

Could be OK to leave as is - that implementation detail is fairly obvious.

Reply:
Do you mean to add some option like `permitBridgeInterfaceOnPodNetwork: true`?
I think this is not required. Unlike bridge this method way more safe.

From the node perspective traffic from another side of veth interface no different from the normal Pod without VM.
The MAC address will always be filtered to the single one used in macvtap interface.

Reply:
I simply mean the current admitter code will reject creating a VMI having a macvtap interface meant for pod network - afaiu,  "allow using macvtap mode to bind pod networking" means relaxing that rule, and allow macvtap interface to be created also for the pod network **if** the feature gate is enabled.